File: android/app/src/main/java/com/tacmap/localization/DisplayFormat.kt
package com.tacmap.localization

import java.math.RoundingMode
import java.text.DateFormat
import java.text.NumberFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/** Presentation only. Never use for coordinates, storage, signatures or exports. */
object DisplayFormat {
    fun locale(languageTag: String?, device: Locale = Locale.getDefault()): Locale {
        if (languageTag == null) return device
        return Locale.Builder().setLanguageTag(languageTag).apply {
            if (device.country.isNotEmpty()) setRegion(device.country)
        }.build()
    }

    val currentLocale: Locale get() = locale(
        AppLanguage.selection.takeUnless { it == SupportedLanguage.SYSTEM }?.tag
    )

    fun number(value: Double, decimals: Int, locale: Locale = currentLocale): String {
        if (!value.isFinite()) return "—"
        return NumberFormat.getNumberInstance(locale).apply {
            isGroupingUsed = false
            minimumFractionDigits = decimals
            maximumFractionDigits = decimals
            roundingMode = RoundingMode.HALF_EVEN
        }.format(value)
    }

    fun distance(metres: Double, locale: Locale = currentLocale): String = when {
        metres < 1000 -> number(metres, 0, locale) + " m"
        else -> number(metres / 1000, if (metres < 100_000) 2 else 0, locale) + " km"
    }

    fun area(squareMetres: Double, locale: Locale = currentLocale): String = when {
        squareMetres < 10_000 -> number(squareMetres, 0, locale) + " m²"
        squareMetres < 1_000_000 -> number(squareMetres / 10_000, 2, locale) + " ha"
        else -> number(squareMetres / 1_000_000, 2, locale) + " km²"
    }

    fun time(date: Date, locale: Locale = currentLocale, timeZone: TimeZone = TimeZone.getDefault()): String =
        DateFormat.getTimeInstance(DateFormat.SHORT, locale).apply { this.timeZone = timeZone }.format(date)
}

File: ios/TacticalMaps/Util/DisplayFormat.swift
import Foundation

/// Presentation only. Never use for coordinates, storage, signatures or exports.
enum DisplayFormat {
    static func locale(languageTag: String?, device: Locale = .current) -> Locale {
        guard let languageTag else { return device }
        var components = Locale.Components(identifier: languageTag)
        if let region = device.region { components.languageComponents.region = region }
        return Locale(components: components)
    }

    static var currentLocale: Locale {
        let choice = AppLanguage.shared.selection
        return locale(languageTag: choice == .system ? nil : choice.rawValue)
    }

    static func number(_ value: Double, decimals: Int, locale: Locale = currentLocale) -> String {
        guard value.isFinite else { return "—" }
        let formatter = NumberFormatter()
        formatter.locale = locale
        formatter.numberStyle = .decimal
        formatter.usesGroupingSeparator = false
        formatter.minimumFractionDigits = decimals
        formatter.maximumFractionDigits = decimals
        formatter.roundingMode = .halfEven
        return formatter.string(from: NSNumber(value: value)) ?? "—"
    }

    static func distance(_ metres: Double, locale: Locale = currentLocale) -> String {
        if metres < 1000 { return number(metres, decimals: 0, locale: locale) + " m" }
        return number(metres / 1000, decimals: metres < 100_000 ? 2 : 0, locale: locale) + " km"
    }

    static func area(_ squareMetres: Double, locale: Locale = currentLocale) -> String {
        if squareMetres < 10_000 { return number(squareMetres, decimals: 0, locale: locale) + " m²" }
        if squareMetres < 1_000_000 { return number(squareMetres / 10_000, decimals: 2, locale: locale) + " ha" }
        return number(squareMetres / 1_000_000, decimals: 2, locale: locale) + " km²"
    }

    static func time(_ date: Date, locale: Locale = currentLocale, timeZone: TimeZone = .current) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = timeZone
        formatter.dateStyle = .none
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }
}
